### ISSUE TITLE
fix AttributeError: 'Axes' object has no property 'axisbg'

### ISSUE THREAD

**Issue by ArtyomKaltovich** [biobakery/MetaPhlAn#87]
Hello.

When I try to create heatmap it fails with `AttributeError: 'Axes' object has no property 'axisbg'`

**Comment by fbeghini** [comment]
The `metaphlan_hclust_heatmap.py` script is deprecated, I'd advise you for using [hclust2](https://github.com/SegataLab/hclust2) for heatmap visualization. The script is removed in MetaPhlAn 3.0

**Comment by charlottecc** [comment]
Hi, 

I'm having trouble getting the reccomended Hclust2.py command to work. I installed metaphlan3 via Conda, would you recommend creating a new environment for Hclust2 as it requires python 2.7? 

Thanks!

**Comment by fbeghini** [comment]
Have you installed version 1.0.0? The latest one is compatible with python 3.7

**Comment by charlottecc** [comment]
I tried to install the latest version with `conda install -c bioconda hclust2 ` in my Conda environment which I installed MetaPhlAn3 in. 

I get this error message : 
`Found conflicts! Looking for incompatible packages. 
failed                                                                                                  
UnsatisfiableError: The following specifications were found
to be incompatible with the existing python installation in your environment:
Specifications:
 hclust2 -> python[version='2.7.*|<3|>=2.7,<2.8.0a0']
Your python: python=3.7
If python is on the left-most side of the chain, that's the version you've asked for.
When python appears to the right, that indicates that the thing on the left is somehow
not available for the python version you are constrained to. Note that conda will not
change your python version to a different minor version unless you explicitly specify
that. `

**Comment by fbeghini** [comment]
It's trying to install 0.99 instead of 1.0.0. Have you configured Bioconda
according to https://bioconda.github.io/user/install.html#set-up-channels?
Can you try installing hclust2 in a new environment with just python 3.7
installed inside and then run `conda install hclust2=1.0.0`?

----

Francesco Beghini

PhD Student


Lab. of Computational Metagenomics

Department of Cellular, Computational and Integrative Biology - CIBIO

University of Trento
Via Sommarive 9, 38123 Trento, Italy

Il Mar 7 Lug 2020, 22:27 charlottecc <notifications@github.com> ha scritto:

**Comment by charlottecc** [comment]
I have configured Bioconda, running the commands on the page you linked. 

I tried `conda install hclust2=1.0.0` in my metaphlan3 env and a new python3.7 env and I still get the error: 

`UnsatisfiableError: The following specifications were found
to be incompatible with the existing python installation in your environment:
Specifications:
  hclust2=1.0.0 -> python[version='<3']
Your python: python=3.7`

Thank you for your help!

**Comment by fbeghini** [comment]
It looks like that till yesterday night the new build was still not available on bioconda. Now it's available but the bioconda autobump kept the python 2.7 requirement. I've updated the requirements and in a few hours it should be available again. Meanwhile, 1.0.0 is available, but only for python 2.7.

**Comment by charlottecc** [comment]
Thank you so much for your help. I have now managed to install Hclust2 in my metaphlan3 Conda environment. 

Although I am now getting error messages when I run the recommended code : 

`hclust2.py \
  -i statemerged_abundance_table.txt \
  -o HMP.sqrt_scale.png \
  --skip_rows 1 \
  --ftop 50 \
  --f_dist_f correlation \
  --s_dist_f braycurtis \
  --cell_aspect_ratio 9 \
  -s --fperc 99 \
  --flabel_size 4 \
  --metadata_rows 2,3,4 \
  --legend_file HMP.sqrt_scale.legend.png \
  --max_flabel_len 100 \
  --metadata_height 0.075 \
  --minv 0.01 \
  --no_slabels \
  --dpi 300 \
  --slinkage complete`

Error: 
`Traceback (most recent call last):
  File "/pub38/cchong/miniconda2/envs/metaphlan3/bin/hclust2.py", line 825, in <module>
    hclust2_main()
  File "/pub38/cchong/miniconda2/envs/metaphlan3/bin/hclust2.py", line 784, in hclust2_main
    dm = DataMatrix( args.inp, args )
  File "/pub38/cchong/miniconda2/envs/metaphlan3/bin/hclust2.py", line 174, in __init__
    select( self.args.fperc, self.args.ftop )
  File "/pub38/cchong/miniconda2/envs/metaphlan3/bin/hclust2.py", line 159, in select
    self.table['perc'] = self.table.apply(lambda x: stats.scoreatpercentile(x,perc),axis=1)
  File "/pub38/cchong/miniconda2/envs/metaphlan3/lib/python3.7/site-packages/pandas/core/frame.py", line 6878, in apply
    return op.get_result()
  File "/pub38/cchong/miniconda2/envs/metaphlan3/lib/python3.7/site-packages/pandas/core/apply.py", line 186, in get_result
    return self.apply_standard()
  File "/pub38/cchong/miniconda2/envs/metaphlan3/lib/python3.7/site-packages/pandas/core/apply.py", line 296, in apply_standard
    values, self.f, axis=self.axis, dummy=dummy, labels=labels
  File "pandas/_libs/reduction.pyx", line 620, in pandas._libs.reduction.compute_reduction
  File "pandas/_libs/reduction.pyx", line 128, in pandas._libs.reduction.Reducer.get_result
  File "/pub38/cchong/miniconda2/envs/metaphlan3/bin/hclust2.py", line 159, in <lambda>
    self.table['perc'] = self.table.apply(lambda x: stats.scoreatpercentile(x,perc),axis=1)
  File "/pub38/cchong/miniconda2/envs/metaphlan3/lib/python3.7/site-packages/scipy/stats/stats.py", line 1891, in scoreatpercentile
    sorted_ = np.sort(a, axis=axis)
  File "<__array_function__ internals>", line 6, in sort
  File "/pub38/cchong/miniconda2/envs/metaphlan3/lib/python3.7/site-packages/numpy/core/fromnumeric.py", line 970, in sort
    a.sort(axis=axis, kind=kind, order=order)
TypeError: '<' not supported between instances of 'float' and 'str'`

Could you offer any advice? 

Thank you in advance!!

**Comment by fbeghini** [comment]
Can you upload here the `statemerged_abundance_table.txt` file? It seems you have a numeric field with a `<` symbol inside of it

**Comment by charlottecc** [comment]
I can't seem to find the `<` but here is the file.

[statemerged_abundance_table.txt](https://github.com/biobakery/MetaPhlAn/files/4896243/statemerged_abundance_table.txt)

Thanks!

**Comment by fbeghini** [comment]
Hi @charlottecc,
I gave a look at the command you used and the input data: the data is OK, the command was the problem, probably you copied-pasted directly. 
Since you don't have metadata in your input file, the `--metadata_rows 2,3,4` is not needed, this led to use the header + some lines with the relative abundances as metadata and raised the error you reported.
Also, there's no need to skip the first line (`--skip_rows 1`), starting from `hclust2` version 1.0.0 the commented header lines are ignored.

I leave you here below the correct command line I used

`hclust2.py -i statemerged_abundance_table.txt -o HMP.sqrt_scale.png --ftop 50 --f_dist_f correlation --s_dist_f braycurtis --cell_aspect_ratio 9 -s --fperc 99 --flabel_size 4 --legend_file HMP.sqrt_scale.legend.png --max_flabel_len 100 --metadata_height 0.075 --minv 0.01 --no_slabels --dpi 300 --slinkage complete`

**Comment by charlottecc** [comment]
Thank you for getting back to me @fbeghini I will use the new command. Out of interest is there a way to add my metadata in?

**Comment by fbeghini** [comment]
You need to add for each sample in the input file a row for each metadata
and then specify the row numbers in the command line. For an example, have
a look at the example [here](
https://github.com/SegataLab/hclust2/blob/master/examples/HMP-MetaPhlAn/HMP.species.txt
)

----

Francesco Beghini

PhD Student


Lab. of Computational Metagenomics

Department of Cellular, Computational and Integrative Biology - CIBIO

University of Trento
Via Sommarive 9, 38123 Trento, Italy

Il Ven 10 Lug 2020, 18:59 charlottecc <notifications@github.com> ha scritto:

**Comment by charlottecc** [comment]
Thank you for all your help!!!

**Comment by charlottecc** [comment]
Hi, 

I'm having another issue now when I try to run hclust2. What version of matplotlib should be installed? The command above is running but the resulting heatmap is basically just the taxon labels and I get the error:

`/pub38/cchong/miniconda2/envs/metaphlan3/bin/hclust2.py:74: MatplotlibDeprecationWarning: 
The iterable function was deprecated in Matplotlib 3.1 and will be removed in 3.3. Use np.iterable instead.
  if matplotlib.cbook.iterable(value):`

Any idea what the issue is? 

Thanks!

**Comment by fbeghini** [comment]
The latest available (3.2.2) is OK, the reason why you see all labels and the heatmap is kinda a strip is because you have to trim the species name, I'd suggest to drop everything before the species name using sed (see https://github.com/biobakery/biobakery/wiki/metaphlan2#create-a-heatmap-with-hclust2)

**Comment by charlottecc** [comment]
Thank you for getting back to me! I adjusted some of the parameters and got my heat maps looking great! 

I'm now however having an issue with export2graphlan.py. I installed this and graphlan in another Conda env as they require python 2.7.

When I run:
`export2graphlan.py --skip_rows 1 -i statemerged_abundance_table_reformatted.txt --tree merged_abundance.tree.txt --annotation merged_abundance.annot.txt --most_abundant 100 --abundance_threshold 1 --least_biomarkers 10 --annotations 5,6 --external_annotations 7 --min_clade_size 1`

I get the error:
`Traceback (most recent call last):
  File "/pub38/cchong/miniconda2/envs/graphlan/bin/export2graphlan.py", line 770, in <module>
    main()
  File "/pub38/cchong/miniconda2/envs/graphlan/bin/export2graphlan.py", line 599, in main
    biomarkers = get_biomarkes(abundant, args.least_biomarkers)
  File "/pub38/cchong/miniconda2/envs/graphlan/bin/export2graphlan.py", line 381, in get_biomarkes
    while lvl < len(max(cc)):
ValueError: max() arg is an empty sequence`

Any ideas what's going wrong? 

Thank you so much for all your help!!
